### ISSUE TITLE
[RFC] Allow pipelining data processing and ML computations

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#16852]
The problem: currently, several use cases for combining datasets with ML inference and training desire overlapping preprocessing and ML computation. Here are three scenarios that have come up:

## Use cases
**Batch inference**: The user wants to preprocess a large number of files from S3, and then use GPU actors to perform batch inference on it. Since the preprocessing stage can take a long time, we want to overlap that with the GPU computation to avoid idling GPUs in the cluster.

**Streaming data into a training job**: The user wants to stream data into a distributed training job (split the data into `N` iterators to feed into `N` training actors). Similar to the previous use case, the goal is to minimize idle GPUs.

**Shuffling a dataset repeatedly per epoch and feeding it into a training job**: The user wants to do a full distributed shuffle of each epoch's data (each epoch is over the same dataset). To minimize latency, we should kick off the shuffle for the next epoch while training over the current epoch.

## How does pipelining help?

Without pipelining, specialized resources like GPUs can be idle for long periods of time:
```
data loading..................
                              preprocessing....
                                               infer.........
-------------------------------------------------------------------> time
|---------------------------------------------|
                 GPU Idle
```

Pipelining can improve GPU utilization and reduce the overall execution time:
```
load (1/3) load (2/3) load (3/3)
           prep (1/3) prep (2/3) prep (3/3)
                      infer (1/3) infer (2/3) infer (3/3)
----------------------------------------------------------> time
|--------------------|
      GPU Idle
```

## Implementation options

There are a couple high-level options:

Option 1: Support pipelined/streaming execution semantics natively _within_ datasets. This is challenging in terms of implementation since we would have to implement backpressure, etc.
 
Option 2: Add higher-level pipeline functionality _on top_ of datasets. This would also address the above use cases, while decoupling pipelining from the dataset implementation.

## Option 2 API

We could implement Option 2 by adding `.pipeline()` and `.repeat()` operators to `Dataset` as follows. This is similar to the approach taken by Spark DStreams, but is defining a pipeline over a finite Dataset instead of an infinite stream.

```
class Dataset(Generic[T]):
    # convert dataset into a pipeline by splitting it into a series of smaller datasets
    def pipeline(self, length: int = 20) -> DatasetPipeline[T]

    # convert dataset into a pipeline by repeating it
    def repeat(self, stop_after: int = None) -> DatasetPipeline[T]

class DatasetPipeline(Generic[T]):
    # pipelined versions of Dataset methods
    def map(self, fn: Callable[[T], [U], compute="tasks", ...) -> DatasetPipeline[T]
    [flatmap, filter, iter_rows, split, etc...]

    # iterate over the evaluated datasets in this pipeline
    def iter_datasets() -> Iterator[Dataset[T]]
```

Unlike Dataset, DatasetPipeline is _lazy_. No computation occurs unless elements are pulled from the pipeline via methods like write_json() or iter_datasets().

## Examples

The **batch inference** case:

```
ray.data.read_binary_files(dir) \
    .pipeline(length=10) \
    .map(preprocess) \
    .map_batches(batch_infer, num_gpus=1, compute="actors", batch_size=256) \
    .write_json("s3://my-bucket/results")
```

The **streaming ingest** case:

```
splits = ray.data.read_parquet(dir).pipeline(length=50).split(NUM_WORKERS)

@ray.remote
class Worker:
    def __init__(self, split, rank):
        ...

    def train(self):
        for s in self.split.to_torch():
           train(s)

workers = [Worker.remote(s, i) for i, s in enumerate(splits)]
ray.get([workers.train.remote() for w in workers])
```

The **per-epoch global shuffle** case:

```
trainer = HorovodTrainer(NUM_WORKERS)

splits = ray.data.read_parquet(dir) \
    .repeat() \
    .global_shuffle() \
    .split(len(trainer.workers), locality_hints=trainer.workers)

trainer.train(splits)
```

**Comment by zhe-thoughts** [comment]
Thanks @ericl ! Are there any advantages of option 1 (native on Dataset)?

**Comment by ericl** [comment]
You potentially get finer-grained pipelining with option (1), which could reduce a bit of the startup overhead of the pipeline. I think this is the main advantage I can think of (same argument of "real streaming" vs DStreams).

**Comment by crypdick** [comment]
cc @mushuskat
